pymeasure/display/widgets/fileinput_widget.py

```diff
@@ -0,0 +1,153 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2023 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import logging
+
+from ..Qt import QtCore, QtWidgets
+from .filename_widget import FilenameLineEdit
+from .directory_widget import DirectoryLineEdit
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+class FileInputWidget(QtWidgets.QWidget):
+    _extensions = ["csv", "txt"]
+
+    def __init__(self, filename_input=True, directory_input=True, parent=None):
```

Comment:
You are right it is a bit of an uggly construction right now. The MixIn classes is an option, but I think you are right that is is more complicated.

When rethinking this, I _think_ that it is not really necessary to be able to select only the filename or directory input field. So now I went for:
- Just a single class, no option to remove one of the fields. So also only a single boolean toggle in the main-window `enable_file_input` instead of the two filename and directory toggles.
- I did add a flag `filename_fixed` that allows for 'freezing' the filename. In that case, the filename field will be disabled, but still present; such that you clearly see where the file will be saved, but are not able to change it.

What do you think?

Reply:
I like it.